File: go/0474M-一和零.go
/*
 * @Author: SukiEva
 * @Date: 2021-12-01 09:34:48
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

func findMaxForm(strs []string, m int, n int) int {
	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}
	for _, str := range strs {
		one, zero := 0, 0
		for _, char := range str {
			if char == '0' {
				zero++
			} else {
				one++
			}
		}
		for i := m; i >= zero; i-- {
			for j := n; j >= one; j-- {
				dp[i][j] = max(dp[i][j], dp[i-zero][j-one]+1)
			}
		}
	}
	return dp[m][n]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: cpp/0474M-一和零.cpp
/*
 * @Author: SukiEva
 * @Date: 2022-01-10 13:25:43
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        vector<vector<int>> dp(m + 1, vector<int>(n + 1));

        for (int i = 0; i < strs.size(); i++) {
            int zero = 0, one = 0;
            for (char c : strs[i]) {
                if (c == '1')
                    one++;
                else
                    zero++;
            }
            for (int j = m; j >= zero; j--) {
                for (int k = n; k >= one; k--) {
                    dp[j][k] = max(dp[j][k], dp[j - zero][k - one] + 1);
                }
            }
        }
        return dp[m][n];
    }
};